core/distributed/matrix.cpp

```diff
@@ -0,0 +1,440 @@
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/distributed/matrix.hpp>
+
+
+#include <ginkgo/core/distributed/vector.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/distributed/matrix_kernels.hpp"
+
+
+namespace gko {
+namespace distributed {
+namespace matrix {
+namespace {
+
+
+GKO_REGISTER_OPERATION(build_diag_offdiag,
+                       distributed_matrix::build_diag_offdiag);
+
+
+}  // namespace
+}  // namespace matrix
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm)
+    : Matrix(exec, comm, with_matrix_type<gko::matrix::Csr>())
+{}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    const LinOp* inner_matrix_type)
+    : Matrix(exec, comm, inner_matrix_type, inner_matrix_type)
+{}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+Matrix<ValueType, LocalIndexType, GlobalIndexType>::Matrix(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    const LinOp* inner_matrix_type, const LinOp* ghost_matrix_type)
+    : EnableLinOp<
+          Matrix<value_type, local_index_type, global_index_type>>{exec},
+      DistributedBase{comm},
+      send_offsets_(comm.size() + 1),
+      send_sizes_(comm.size()),
+      recv_offsets_(comm.size() + 1),
+      recv_sizes_(comm.size()),
+      gather_idxs_{exec},
+      local_to_global_ghost_{exec},
+      one_scalar_{},
+      diag_mtx_{inner_matrix_type->clone(exec)},
+      offdiag_mtx_{ghost_matrix_type->clone(exec)}
```

Comment:
Perhaps local/neighbor? But I don't think there is a nice application agnostic choice for that out there.

PS: I like inner/ghost because both are the same length.